### ISSUE TITLE
Unknown rows returning different kWh & co2e totals when using different time periods

### ISSUE THREAD

**Issue by jhowlett-scottlogic** [cloud-carbon-footprint/cloud-carbon-footprint#1200]
### Preflight Checklist

- [x] I have read the [Contributing Guidelines](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/trunk/CONTRIBUTING.md) for this project.
- [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

- **Operating System:**
  -  Windows 10 
- **Browser:**
  - Chrome however not browser specific 

### Expected Behavior

See the same kWh & co2e totals for data points when using different time periods for data capture

### Actual Behavior

Depending on the time period stated in client/.env using `REACT_APP_START_DATE` & `REACT_APP_END_DATE` you get different kWh & co2e totals for data points. Note: This isn't specific to a `REACT_APP_GROUP_BY` value.

### To Reproduce

Have AWS (I know it occurs with AWS, not sure about the others) data that will enter into method: `CostAndUsageReports.ts:getEstimateForUnknownUsage()` (happens when `this.getFootprintEstimateFromUsageRow()` returns `undefined`) . This method returns different kWh & co2 estimates for each data point depending on the time period used.

I have also been using data that sets `serviceAndUsageUnit` to undefined and therefore doesnt get into the first if-block in `UnknownEstimator.ts:estimateKilowattHours()` but instead gets into the second by setting `totalForUsageUnit`. Therefore i know this behaviour occurs in the second if-block but i am not sure if it happens in the first if-block too.

I have reproduced this by using group by of day. Then a period of 1 day and looking at the cache, then using 2 days (day interested in + day after) and then looking at the cache total for the day interested in. They are different if they include rows that are unknown.

### Additional Information

I have seen this issue: https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/issues/839 however it only talks about co2e and no reference to altering time periods. May be linked but not certain

© 2021 Thoughtworks, Inc.

**Comment by jhowlett-scottlogic** [comment]
From further investigation, it looks like this is due to the estimation of unknown rows using the total kWh over entire time period instead of the total for that data point period. Therefore if you change to time period, the totals change.

UknownEstimator.ts

    if (totalForUsageUnit)
      const x =  (totalForUsageUnit.kilowattHours /
          totalForUsageUnit[this.estimateKilowattHoursBy]) *
        unknownUsage[this.estimateKilowattHoursBy]


Both `totalForUsageUnit.kilowattHours` & `totalForUsageUnit[this.estimateKilowattHoursBy])` change depending on the time period used and so the estimated values returned are different. 

To confirm this is the root of the problem, if instead you only use the totals for a specific day it can be confirmed that it is the differing totals causing this behavior. To do this, in `FootprintEstimate.ts:accumulateKilowattHours()` you can add an if statement to only accumulate values for a specific day using `  if(billingDataRow.timestamp.toDateString() == new Date('{day you are testing}').toDateString()) {`.

Suggested solution is to not use the total for the whole time period but instead use the total for that day. Do not want to do it per data point period (day/week/month) because you would get the same wrong behavior when you change the time interval

**Comment by jhowlett-scottlogic** [comment]
I have been able to create a test that simulates this behaviour. In `UnknownEstimator.test.ts` you can add this test:

```
    it('does estimates for unknown usage type for AWS, based on cost, with different time period', () => {
      const service = 'testService'
      const usageUnit = 'testUsageUnit'
      const input = [
        {
          timestamp: new Date('2021-01-01'),
          cost: 1000,
          usageUnit: usageUnit,
          service: service,
        },
        {
          timestamp: new Date('2021-01-02'),    // This is the second data point that shouldn't be used.
          cost: 1234,   
          usageUnit: usageUnit,
          service: service,
        },
      ]
      const awsUsEast1Region = 'us-east-1'
      const awsEmissionsFactors = {
        [awsUsEast1Region]: 0.000415755,
      }

      const awsConstants = {
        kilowattHoursByServiceAndUsageUnit: {
          [service]: {
            [usageUnit]: {
              cost: 50,
              kilowattHours: 20,
            },
          },
        },
      }
      const result = new UnknownEstimator(EstimateUnknownUsageBy.COST).estimate(
        input,
        awsUsEast1Region,
        awsEmissionsFactors,
        awsConstants,
      )

      expect(result).toEqual([
        {
          co2e: 0.166302,
          timestamp: new Date('2021-01-01T00:00:00.000Z'),
          kilowattHours: 400,
          usesAverageCPUConstant: false,
        },
      ])
    })
```

It is a copy of the first test in the file with an extra data point added for a different date. 

The existing test simulates using a time period of 2021-01-01 - 2021-01-01 and this new test simulates using a time period of 2021-01-01 - 2021-01-02 and checks if the data for date 2021-01-01 is the same. However this new test fails

**Comment by ccasher** [comment]
Hi @jhowlett-scottlogic thank you for raising this issue and providing the detailed context! We are actually aware of this behavior and have had some internal discussion around it.

I think you have a solid idea by averaging the unknown estimations per day, but one immediate potential drawback that comes to mind is that when calculating the unknowns over a longer period of time, there will be a higher degree of accuracy due to the larger sample size of known estimations. So that means this would provide slightly less accurate unknown estimates because it would only consider the daily values.

Therefore, there becomes a trade off of accuracy vs consistency. One thought is to even add this as an optional configuration for users to decide for themselves.

Interested to hear your thoughts!

**Comment by jhowlett-scottlogic** [comment]
Hi @ccasher, thanks for getting back to me on this issue. 

I agree entirely with the reasoning for an against using the total rows for estimations. 

I see the most common use case for the calculator to be:  use `seed-cache-file` get x months of data in the cache. Set the `REACT_APP_START_DATE` to the date you used in the `seed-cache-file` command and do not set a `REACT_APP_END_DATE` so that everyday the tool retrieves the previous day's data. The estimated rows would slowly get more accurate as the data used to calculate the unknown rows becomes larger. For this case the use of the total over the time period is best

I am trying to think of a common use case for the day estimations, my scenario is really the only one i can think of. I used 12 months for the `seed-cache-file` command and retrieved the results and did some analysis on them. I then realized i wanted another 6 months so deleted the cache and did `seed-cache-file` for 18 months expecting the data for the days already retrieved to be identical. They were wildly different, trends were the same but totals were a factor out, and meant that i did not know what was the most accurate data (now i understand the 18 month is). It meant i needed to redo all my analysis using the updated values. So for a use that constantly changes time periods or there are two instances of the tool running (possibly one for 1 months data and 1 for last years data on separate dashboards?) you would want the estimations based on the day totals. 

From this it seems the former scenario will be the most common and so the estimation method using the overall total should definitely be kept and the addition of the day estimation is really just a nice to have. 

One thing that would prevent this coming up again would be to include a section in the methodology that talks about the estimation of unknown rows as this i feel is a crucial part of the overall methodology.

**Comment by ccasher** [comment]
Thanks for your reply @jhowlett-scottlogic. Just to clarify - wrt the former scenario, the cached data should not be affecting the newly fetched unknown estimation calculation. The unknown estimates are calculated during the application run time, only using the known estimates stored in memory. We have documentation on this process [here](https://www.cloudcarbonfootprint.org/docs/classifying-usage-types#handling-unknown-usage-types), but may be worth considering updating to reflect the call out on varying totals based on time frames.

I do agree that there are likely scenarios where both types of unknown estimation calculations (each day vs overall total) could be useful for users. We can plan to add this to our project board. If you might be interested in picking up this task to implement an optional configuration and submitting a pull request, that would be very welcome!

**Comment by jhowlett-scottlogic** [comment]
@ccasher I did miss understand the process in the former scenario a bit then.

So just to confirm, every day after the initial cache population, when the calculator is started so that it can retrieve the new usage data from the CSPs, the next day's unknown rows uses only that day's totals for estimation. 

That is a bit of a shame as it would be great for the tool to slowly become more accurate with time by using the total over the whole time period.

Maybe, if the tool could use the cache directly for estimations rather than the freshly retrieved data you could specific the time period you want to use when estimating unknown rows?

**Comment by 4upz** [comment]
@jhowlett-scottlogic I agree in that we can approach this in two ways:

1. Providing users the option to specify via config/parameter whether they want the coefficients used for the unknown estimator to be across the request range or on a per-day basis
2. Providing an even more robust support for persisting the unknown estimators' coefficient for all data collected. This could allow for it to become more accurate the more data is collected and time ranges are requested. 

I think there is room to do both, and believe it should be up to the user to weigh the tradeoffs and decide which method best fits their use case. The second option should be very much possible using a separate MongoDB collection or local storage. With the first option, we need to be aware that a daily average would be less accurate when estimates are grouped by a range other than a "day" (i.e. week, month, quarter, etc.). It may need to be based on the chosen grouping method instead. Regardless, we'll add both approaches to our backlog and try to explore them.

**Comment by jhowlett-scottlogic** [comment]
@4upz I agree with both approaches. They are very much 2 separate pieces of work, the first is the addition of a new estimation approach and the second is an improvement on the current estimation approach that can be made when using a separate MongoDB collection or local storage.


 Just one thing to note, if you use the time interval in the config (i.e. week, month, quarter, etc.) rather than always a day you will get different results for different time intervals. Causing the same problem as currently exists with the 'whole time period' approach and changing time periods. That is why I have suggested always using a day, if this isn't plausible then it would be worth adding a disclaimer stating that if you change time interval, you will get different totals for estimated rows and therefore a different overall total.
